### PR TITLE
change Saved Game feature 

### PR DESCRIPTION
**Pull request by frg2089**
some suggestions:
- Use [System.IO.Path](//docs.microsoft.com/dotnet/api/system.io.path) Member and **NOT** use character `'/'` or `'\\'` to cross platform.
- Use Properties instead of Methods.
- Make [.editorconfig](//docs.microsoft.com/visualstudio/ide/create-portable-custom-editor-options) file to constraint the code style.
- Use [nuget](//nuget.org) package instead of reference library.
- Use [Multilingual app toolkit](//docs.microsoft.com/windows/uwp/design/globalizing/use-mat) (well, it will be hard.)

### CODE REVIEW

**Comment by SadPencil** [comment]
```cs
public static string GetSaveGameName(Stream stream)
        {
            var buffer = new byte[0x40];
```
Why would you set the limitation of the length of the saving game name while the game executable supports a much longer name?

**Comment by SadPencil** [comment]
Apart from the limitation of the length, your new name loading logic brings other problems:

In the document of `Encoding.GetString Method`:
```
Exceptions
ArgumentException

The byte array contains invalid Unicode code points.
```

Suspose the name is shorter than `0x40`, it may causes the `ArgumentException`.

So, I strongly disagree with your new logic for loading the name of saved games. It added unnecessary limitations as well as unexpected exceptions. The current one in the repo is nicer.

**Comment by frg2089** [comment]
> ```cs
> public static string GetSaveGameName(Stream stream)
>         {
>             var buffer = new byte[0x40];
> ```
> 
> Why would you set the limitation of the length of the saving game name while the game executable supports a much longer name?

sav file have a fixed block to save name.
at 0x09c0 and length is 0x40.
Actually, I don't know the other at 0x08d0 and there was no length limit.

**Comment by AutoGavy** [comment]
你们两个老外可以说中文不

**Comment by SadPencil** [comment]
> 
> 
> > ```cs
> > public static string GetSaveGameName(Stream stream)
> >         {
> >             var buffer = new byte[0x40];
> > ```
> > 
> > 
> > Why would you set the limitation of the length of the saving game name while the game executable supports a much longer name?
> 
> sav file have a fixed block to save name.
> at 0x09c0 and length is 0x40.
> Actually, I don't know the other at 0x08d0 and there was no length limit.

Which means, your code brings a stronger assumption on the `.sav` file: The name block is fixed to `0x40` and when a null character (0x00 0x00) is set, the rest bytes in the name block must also be `0x00` ( or any other bytes that doesn't raise `ArgumentException` ).

Well, in my opinion, a weaker assumption is better than a stronger one. Or, specify the assumption in the comment.

**Comment by SadPencil** [comment]
Your latest code brings a new issue here:
```cs
            buffer = new byte[(BitConverter.ToInt32(buffer, 0) - 1) << 1];
```
You can't assume whether the computer is little-endian or big-endian, on which the behavior of BitConverter.ToInt32 depends.

**Comment by SadPencil** [comment]
Actually, you should use `BitConverter.IsLittleEndian` to determine whether a reverse is needed, instead of picking a minimum value.
And only use English for comments please.

**Comment by SadPencil** [comment]
Method `SavedGameManager.GetLength` make no sense.
From my view, the administrators in this repo only accept codes that is beautifully written. A code with such a method is definitely not pretty. 

Or you may wait for the suggestion about coding style from the administrators.

File: ClientCore/SavedGameManager.cs [hunk 1]
```diff
@@ -76,10 +76,7 @@ public static List<string> GetSaveGameTimestamps()
 
         public static bool AreSavedGamesAvailable()
         {
-            if (Directory.Exists(SaveGameDirectoryPath))
-                return true;
-
-            return false;
+            return Directory.Exists(SaveGameDirectoryPath);
```
Comment (Metadorius) [inline ClientCore/SavedGameManager.cs:79]: Should use expression body here.

**Comment by Metadorius** [comment]
> Use System.IO.Path Member and NOT use character '/' or '\\' to cross platform.
What's the reason for this? `/` is way simpler to use and is supported by bot Windows and Unix. Also if you do it like this you have to change all occurences of slash also (again, for what?)

**Comment by frg2089** [comment]
> > Use System.IO.Path Member and NOT use character '/' or '\' to cross platform.
> 
> What's the reason for this? `/` is way simpler to use and is supported by bot Windows and Unix. Also if you do it like this you have to change all occurences of slash also (again, for what?)

Used '/' is wrong. it should not be used in windows. Although it works, but it is not native to windows. New code should pay attention to this.
And it's only a suggestion, you don't have to take it.
[file-path-formats#canonicalize-separators](//docs.microsoft.com/dotnet/standard/io/file-path-formats#canonicalize-separators)
[System.IO.Path](//docs.microsoft.com/dotnet/api/system.io.path)

File: ClientCore/SavedGameManager.cs (hunk 1 above)
Comment (frg2089) [reply ClientCore/SavedGameManager.cs:79]: I'll fix it.
And you can make an .editorconfig file to constraint it.

**Comment by Zero-Fanker** [comment]
> > > ```cs
> > > public static string GetSaveGameName(Stream stream)
> > >         {
> > >             var buffer = new byte[0x40];
> > > ```
> > > 
> > > 
> > > Why would you set the limitation of the length of the saving game name while the game executable supports a much longer name?
> > 
> > 
> > sav file have a fixed block to save name.
> > at 0x09c0 and length is 0x40.
> > Actually, I don't know the other at 0x08d0 and there was no length limit.
> 
> Which means, your code brings a stronger assumption on the `.sav` file: The name block is fixed to `0x40` and when a null character (0x00 0x00) is set, the rest bytes in the name block must also be `0x00` ( or any other bytes that doesn't raise `ArgumentException` ).
> 
> Well, in my opinion, a weaker assumption is better than a stronger one. Or at least specify the assumption in the comment.

See 'SavegameInformation' as Alexb discovered here :
https://github.com/Ares-Developers/YRpp/blob/master/LoadOptionsClass.h#L95

**Comment by SadPencil** [comment]
> 
> 
> > > > ```cs
> > > > public static string GetSaveGameName(Stream stream)
> > > >         {
> > > >             var buffer = new byte[0x40];
> > > > ```
> > > > 
> > > > 
> > > > Why would you set the limitation of the length of the saving game name while the game executable supports a much longer name?
> > > 
> > > 
> > > sav file have a fixed block to save name.
> > > at 0x09c0 and length is 0x40.
> > > Actually, I don't know the other at 0x08d0 and there was no length limit.
> > 
> > 
> > Which means, your code brings a stronger assumption on the `.sav` file: The name block is fixed to `0x40` and when a null character (0x00 0x00) is set, the rest bytes in the name block must also be `0x00` ( or any other bytes that doesn't raise `ArgumentException` ).
> > Well, in my opinion, a weaker assumption is better than a stronger one. Or at least specify the assumption in the comment.
> 
> See 'SavegameInformation' as Alexb discovered here :
> https://github.com/Ares-Developers/YRpp/blob/master/LoadOptionsClass.h#L95

Good to know. Thanks.

**Comment by SadPencil** [comment]
> 
> 
> > > > ```cs
> > > > public static string GetSaveGameName(Stream stream)
> > > >         {
> > > >             var buffer = new byte[0x40];
> > > > ```
> > > > 
> > > > 
> > > > Why would you set the limitation of the length of the saving game name while the game executable supports a much longer name?
> > > 
> > > 
> > > sav file have a fixed block to save name.
> > > at 0x09c0 and length is 0x40.
> > > Actually, I don't know the other at 0x08d0 and there was no length limit.
> > 
> > 
> > Which means, your code brings a stronger assumption on the `.sav` file: The name block is fixed to `0x40` and when a null character (0x00 0x00) is set, the rest bytes in the name block must also be `0x00` ( or any other bytes that doesn't raise `ArgumentException` ).
> > Well, in my opinion, a weaker assumption is better than a stronger one. Or at least specify the assumption in the comment.
> 
> See 'SavegameInformation' as Alexb discovered here :
> https://github.com/Ares-Developers/YRpp/blob/master/LoadOptionsClass.h#L95

After carefully reviewing the code you provided, the assumption given by @frg2089 at first is still stronger and thus not proper.

On his previous code (which has been deleted by a force push), he writes the code like this:
```
1. Get a Unicode string with the 40 bytes data. // Can raise exception on some bytes that is not UTF-16.
2. Remove any null characters at the end. // Assume that unused bytes are 0x00, which is a strong assumption.
```

However, the behavior of file `LoadOptionsClass.h` is different. The main difference is that, the type of archive name is `FixedString`, which behaves like a Pascal string. The content of unused bytes are **undefined**, which means it can be anything. One can't assume that these unused bytes are '0x00', which is the main problem of his previous code.

**Comment by SadPencil** [comment]
> 
> 
> > > > ```cs
> > > > public static string GetSaveGameName(Stream stream)
> > > >         {
> > > >             var buffer = new byte[0x40];
> > > > ```
> > > > 
> > > > 
> > > > Why would you set the limitation of the length of the saving game name while the game executable supports a much longer name?
> > > 
> > > 
> > > sav file have a fixed block to save name.
> > > at 0x09c0 and length is 0x40.
> > > Actually, I don't know the other at 0x08d0 and there was no length limit.
> > 
> > 
> > Which means, your code brings a stronger assumption on the `.sav` file: The name block is fixed to `0x40` and when a null character (0x00 0x00) is set, the rest bytes in the name block must also be `0x00` ( or any other bytes that doesn't raise `ArgumentException` ).
> > Well, in my opinion, a weaker assumption is better than a stronger one. Or at least specify the assumption in the comment.
> 
> See 'SavegameInformation' as Alexb discovered here :
> https://github.com/Ares-Developers/YRpp/blob/master/LoadOptionsClass.h#L95

According to issue https://github.com/CnCNet/xna-cncnet-client/issues/164, the archive file is OLE Compound Document format.
The offset of the string is NOT ALWAYS at 0x09c0 or 0x08d0. It varies from platforms and programs. One should not assume a constant offset.

**Pull request closed without merging**